### ISSUE TITLE
Issues with HttpClient Cache / ThreadContext

### ISSUE THREAD

**Issue by SpaceCondor** [SAP/cloud-sdk#698]
## Issue Description

While executing requests using the `.tryExecute()` syntax, I get warning regarding a missing thread context.

The code I am trying to execute is the following:

```
var addDocumentRequest = svcl.withServicePath(svclConfig.getServiceLayerPath())
		.createOrders(salesOrderDocument).withHeader("COOKIE", "B1SESSION=" + sessionID)
		.withHeader("Prefer", "return-no-content");

var serviceLayerDestination = DestinationAccessor.getDestination("service_layer").asHttp();


// Attempt to add the document
var addResponse = addDocumentRequest.tryExecute(serviceLayerDestination);
```

I was reading through the following documentation: [here](https://sap.github.io/cloud-sdk/docs/java/features/multi-tenancy/multi-tenancy-thread-context#running-asynchronous-operations) and the issue seems to be related to spawning new threads or running async operations.

Although I am not seeing documentation that seems applicable to my situation. The above should all be happening on the same thread. Only the destination accessor would've been created on another thread, is that the reason?

## Error Message

```
Could not get HttpClient cache.
com.sap.cloud.sdk.cloudplatform.thread.exception.ThreadContextAccessException: No ThreadContext available for thread id=6255.
	at com.sap.cloud.sdk.cloudplatform.thread.ThreadLocalThreadContextFacade.lambda$tryGetCurrentContext$0(ThreadLocalThreadContextFacade.java:39) ~[cloudplatform-core-3.61.0.jar:?]
	at io.vavr.Value.toTry(Value.java:1414) ~[vavr-0.10.4.jar:?]
	at com.sap.cloud.sdk.cloudplatform.thread.ThreadLocalThreadContextFacade.tryGetCurrentContext(ThreadLocalThreadContextFacade.java:37) ~[cloudplatform-core-3.61.0.jar:?]
	at io.vavr.control.Try.flatMapTry(Try.java:490) ~[vavr-0.10.4.jar:?]
	at io.vavr.control.Try.flatMap(Try.java:472) ~[vavr-0.10.4.jar:?]
	at com.sap.cloud.sdk.cloudplatform.thread.ThreadContextAccessor.tryGetCurrentContext(ThreadContextAccessor.java:84) ~[cloudplatform-core-3.61.0.jar:?]
	at com.sap.cloud.sdk.cloudplatform.connectivity.RequestScopedHttpClientCache.getCache(RequestScopedHttpClientCache.java:28) ~[cloudplatform-connectivity-3.61.0.jar:?]
	at com.sap.cloud.sdk.cloudplatform.connectivity.AbstractHttpClientCache.tryGetOrCreateHttpClient(AbstractHttpClientCache.java:78) ~[cloudplatform-connectivity-3.61.0.jar:?]
	at com.sap.cloud.sdk.cloudplatform.connectivity.AbstractHttpClientCache.tryGetHttpClient(AbstractHttpClientCache.java:46) ~[cloudplatform-connectivity-3.61.0.jar:?]
	at com.sap.cloud.sdk.cloudplatform.connectivity.HttpClientAccessor.tryGetHttpClient(HttpClientAccessor.java:153) ~[cloudplatform-connectivity-3.61.0.jar:?]
	at com.sap.cloud.sdk.cloudplatform.connectivity.HttpClientAccessor.getHttpClient(HttpClientAccessor.java:131) ~[cloudplatform-connectivity-3.61.0.jar:?]
	at com.sap.cloud.sdk.datamodel.odatav4.core.CreateRequestBuilder.execute(CreateRequestBuilder.java:127) ~[odata-v4-core-3.61.0.jar:?]
	at com.sap.cloud.sdk.datamodel.odatav4.core.CreateRequestBuilder.execute(CreateRequestBuilder.java:35) ~[odata-v4-core-3.61.0.jar:?]
	at com.sap.cloud.sdk.datamodel.odatav4.core.RequestBuilderExecutable.lambda$tryExecute$7fa567b0$1(RequestBuilderExecutable.java:55) ~[odata-v4-core-3.61.0.jar:?]
	at io.vavr.control.Try.of(Try.java:75) [vavr-0.10.4.jar:?]
	at com.sap.cloud.sdk.datamodel.odatav4.core.RequestBuilderExecutable.tryExecute(RequestBuilderExecutable.java:55) [odata-v4-core-3.61.0.jar:?]
```

**Comment by Johannes-Schneider** [comment]
Hi @SpaceCondor,

thanks for reaching out to us.

The issue you are observing usually results from forgetting the Spring Boot annotations on your application.
You can refer to [our documentation](https://sap.github.io/cloud-sdk/docs/java/guides/cap-sdk-integration#enable-the-component-scan-for-sap-cloud-sdk-package) to see which exact annotations should be added.

Please let me know whether that helps.

Best regards,
Johannes

**Comment by SpaceCondor** [comment]
Hey @Johannes-Schneider, understood. However this isn't a Spring Boot application. Is there a method or approach for non-spring boot applications?

**Comment by Johannes-Schneider** [comment]
Which framework/library are you using instead of Spring Boot? TomEE?

----

In the end, the easiest way of enabling the Cloud SDK in your application is to let the `com.sap.cloud.sdk.cloudplatform.servlet.RequestAccessorFilter` run.
This will initialize a new `ThreadContext`, which then contains all important information for our Multitenancy features to work.

**Comment by SpaceCondor** [comment]
@Johannes-Schneider No framework currently; I'm using the SDK to generate client code to interact with the SAP Business One Service Layer so a bit different than typical use-cases I assume.

Is there any documentation on using the `RequestAccessorFilter` or examples?

**Comment by Johannes-Schneider** [comment]
That is indeed a little bit of an untypical use-case 😅

Unfortunately, we do not have any public examples for how to enable the `RequestAccessorFilter` manually.

In general, as long as your application uses the Java "Servlet API", I would assume there should be a manual way of adding a new filter to your request processing pipeline.

----

If you cannot use the `RequestAccessorFilter` (e.g. you are not using the "Servlet API"), then you could try initializing a new instance of the `ThreadContext` in your request pipeline like so:

```java
new ThreadContextExecutor()
            .withListeners(new RequestHeaderThreadContextListener(myRequestHeaders))
            .execute(() -> callNextRequestProcessor());
```

That way, the Cloud SDK should be able to find an initialized `ThreadContext` when you are performing your OData requests.

**Comment by SpaceCondor** [comment]
@Johannes-Schneider Understood. I am also not currently using the Servlet API.

So, given my example code in my first comment where would I use the created ThreadContextExecutor? I want to maintain the use of `.tryExecute()`

**Comment by Johannes-Schneider** [comment]
@SpaceCondor 

Depending on whether you want to use the Cloud SDK's multitenancy features, you can try the following approach:

<details>
  <summary>Yes - I want Multitenancy</summary>

```java
// TODO: extract headers from your incoming request to enable Multitenancy
RequestHeaderContainer incomingHeaders;
new ThreadContextExecutor()
    .withListeners(new RequestHeaderThreadContextListener(incomingHeaders))
    .execute(() -> {
        var addDocumentRequest =
            svcl
                .withServicePath(svclConfig.getServiceLayerPath())
                .createOrders(salesOrderDocument)
                .withHeader("COOKIE", "B1SESSION=" + sessionID)
                .withHeader("Prefer", "return-no-content");

        var serviceLayerDestination = DestinationAccessor.getDestination("service_layer").asHttp();

        // Attempt to add the document
        var addResponse = addDocumentRequest.tryExecute(serviceLayerDestination);
    });
```
</details>

<details>
  <summary>No - I don't need Multitenancy</summary>

```java
new ThreadContextExecutor()
    .execute(() -> {
        var addDocumentRequest =
            svcl
                .withServicePath(svclConfig.getServiceLayerPath())
                .createOrders(salesOrderDocument)
                .withHeader("COOKIE", "B1SESSION=" + sessionID)
                .withHeader("Prefer", "return-no-content");

        var serviceLayerDestination = DestinationAccessor.getDestination("service_layer").asHttp();

        // Attempt to add the document
        var addResponse = addDocumentRequest.tryExecute(serviceLayerDestination);
    });
```
</details>

**Comment by SpaceCondor** [comment]
@Johannes-Schneider Perfect, thank you! Appreciate the help, I was struggling a bit with the documentation.

**Comment by Johannes-Schneider** [comment]
@SpaceCondor Glad I was able to help 😊

Could you please point me to the documentation you had trouble with? We are always striving to improve our documentation to help as many consumers as possible, so getting feedback about things that are not quite clear is highly appreciated.

**Comment by SpaceCondor** [comment]
@Johannes-Schneider Since my application doesn't need any multitenancy, I assumed I did not need to propagate/create ThreadContexts.

I know that the use of the SDK for SAP Business One may not be common, so I understand that the documentation isn't necessarily tailored to that. Although, I do wish there were a way to just tell the SDK that multitenancy features are not needed instead of having to wrap _every_ call in a new thread context executor.

**Comment by Johannes-Schneider** [comment]
@SpaceCondor 

Thank you for your feedback! We highly appreciate your opinion and will work on improving our documentation.

---

> Since my application doesn't need any multitenancy, I assumed I did not need to propagate/create ThreadContexts.

I think the "error message" you have shown in your initial report should actually just be a warning, is that right?
So since you are not using any of the Cloud SDK's Multitenancy capabilities, it is indeed safe to ignore this warning and simply **not** wrap your calls inside a `ThreadContextExecutor` to keep your code smaller and, therefore, easier.

Sorry for not mentioning that earlier!

---

Also as a side note, we changed the log level of that exact same message you have observed in one of our latest releases.
Since then, we observed an increasing amount of support incidents similar to yours, as this message is (rightfully IMO) leading to confusion on consumer side.
We are currently working on fixing that issue.
